docs/topics/equality.md

```diff
@@ -1,50 +1,128 @@
 [//]: # (title: Equality)
 
-In Kotlin there are two types of equality:
+In Kotlin, there are two types of equality:
 
-* _Structural_ equality (`==` - a check for `equals()`)
-* _Referential_ equality (`===` - two references point to the same object)
+* _Structural_ equality (`==`) - a check for the `equals()` function
+* _Referential_ equality (`===`) - a check for two references pointing to the same object
 
 ## Structural equality
 
-Structural equality is checked by the `==` operation and its negated counterpart `!=`.
+Structural equality verifies if two objects have the same content or structure. Structural equality is checked by the `==` 
+operation and its negated counterpart `!=`.
 By convention, an expression like `a == b` is translated to:
 
 ```kotlin
 a?.equals(b) ?: (b === null)
 ```
 
-If `a` is not `null`, it calls the `equals(Any?)` function, otherwise (`a` is `null`) it checks that `b`
-is referentially equal to `null`.
+If `a` is not `null`, it calls the `equals(Any?)` function. Otherwise (`a` is `null`), it checks that `b`
+is referentially equal to `null`:
+
+```kotlin
+fun main() {
+    var a = "hello"
+    var b = "hello"
+    var c = null
+    var d = null
+    var e = d
+
+    println(a == b)
+    // true
+    println(a == c)
+    // false
+    println(c == e)
+    // true
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
 
 Note that there's no point in optimizing your code when comparing to `null` explicitly:
 `a == null` will be automatically translated to `a === null`.
 
+In Kotlin, the `equals()` function is inherited by all classes from the `Any` class. By default, the `equals()` function 
+implements [referential equality](#referential-equality). However, classes in Kotlin can override the `equals()` 
+function to provide a custom equality logic and, in this way, implement structural equality.
+
+Value classes and data classes are two specific types in Kotlin that automatically override `equals()`. 
+Value classes override the `equals()` function to implement structural equality by default.
+
+Similarly, data classes override the `equals()` function to implement structural equality by default. However, if the
+`equals()` function is marked as `final` in the parent class, its behavior remains unchanged.
+
+Distinctly, non-data classes (those not declared with the `data` modifier) do not override the 
+`equals()` function by default. Instead, non-data classes implement referential equality behavior inherited from the `Any` class.
+To implement structural equality, non-data classes require a custom equality logic to override the `equals()` function.
```

Comment:
Just though of an idea to give also a recommendation to override hashCode function if you're overriding equals.